# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/find/nsWebBrowserFind.cpp
# Commit: e07e5ea2f75e
# Full Hash: e07e5ea2f75e971967decbbe8f42aff80987fda2
# Author: Jan-Niklas Jaeschke <jjaschke@mozilla.com>
# Date: 2025-02-16 08:52:20
# Regressor Bug: 1945862
# File Overlap Count: 1
# Description:
#   Bug 1945862 - Make `nsFind` respect the end point correctly. r=emilio
#   
#   This patch refactors the find functionality in a few ways:
#   - A C++-only method `FindFromRangeBoundaries()` is introduced,
#     which uses `RangeBoundary`s instead of collapsed `nsRange`s
# ==============================================================================

diff -r 5a1f8c0bb78f -r e07e5ea2f75e toolkit/components/find/nsWebBrowserFind.cpp
--- a/toolkit/components/find/nsWebBrowserFind.cpp	Sat Feb 15 15:19:21 2025 +0000
+++ b/toolkit/components/find/nsWebBrowserFind.cpp	Sat Feb 15 16:02:34 2025 +0000
@@ -371,51 +371,20 @@
       SelectionScrollMode::SyncFlush);
 }
 
-// Adapted from TextServicesDocument::GetDocumentContentRootNode
-nsresult nsWebBrowserFind::GetRootNode(Document* aDoc, Element** aNode) {
-  NS_ENSURE_ARG_POINTER(aDoc);
-  NS_ENSURE_ARG_POINTER(aNode);
-  *aNode = 0;
-
-  if (aDoc->IsHTMLOrXHTML()) {
-    Element* body = aDoc->GetBody();
-    NS_ENSURE_ARG_POINTER(body);
-    NS_ADDREF(*aNode = body);
-    return NS_OK;
-  }
-
-  // For non-HTML documents, the content root node will be the doc element.
-  Element* root = aDoc->GetDocumentElement();
-  NS_ENSURE_ARG_POINTER(root);
-  NS_ADDREF(*aNode = root);
-  return NS_OK;
-}
-
 nsresult nsWebBrowserFind::SetRangeAroundDocument(nsRange* aSearchRange,
                                                   nsRange* aStartPt,
                                                   nsRange* aEndPt,
                                                   Document* aDoc) {
-  RefPtr<Element> bodyContent;
-  nsresult rv = GetRootNode(aDoc, getter_AddRefs(bodyContent));
-  NS_ENSURE_SUCCESS(rv, rv);
-  NS_ENSURE_ARG_POINTER(bodyContent);
-
-  uint32_t childCount = bodyContent->GetChildCount();
-
-  aSearchRange->SetStart(*bodyContent, 0, IgnoreErrors());
-  aSearchRange->SetEnd(*bodyContent, childCount, IgnoreErrors());
+  NS_ENSURE_ARG_POINTER(aDoc);
+  uint32_t childCount = aDoc->GetChildCount();
 
-  if (mFindBackwards) {
-    aStartPt->SetStart(*bodyContent, childCount, IgnoreErrors());
-    aStartPt->SetEnd(*bodyContent, childCount, IgnoreErrors());
-    aEndPt->SetStart(*bodyContent, 0, IgnoreErrors());
-    aEndPt->SetEnd(*bodyContent, 0, IgnoreErrors());
-  } else {
-    aStartPt->SetStart(*bodyContent, 0, IgnoreErrors());
-    aStartPt->SetEnd(*bodyContent, 0, IgnoreErrors());
-    aEndPt->SetStart(*bodyContent, childCount, IgnoreErrors());
-    aEndPt->SetEnd(*bodyContent, childCount, IgnoreErrors());
-  }
+  aSearchRange->SetStart(*aDoc, 0, IgnoreErrors());
+  aSearchRange->SetEnd(*aDoc, childCount, IgnoreErrors());
+
+  aStartPt->SetStart(*aDoc, 0, IgnoreErrors());
+  aStartPt->SetEnd(*aDoc, 0, IgnoreErrors());
+  aEndPt->SetStart(*aDoc, childCount, IgnoreErrors());
+  aEndPt->SetEnd(*aDoc, childCount, IgnoreErrors());
 
   return NS_OK;
 }
@@ -435,13 +404,9 @@
     return SetRangeAroundDocument(aSearchRange, aStartPt, aEndPt, aDoc);
   }
 
-  // Need bodyContent, for the start/end of the document
-  RefPtr<Element> bodyContent;
-  nsresult rv = GetRootNode(aDoc, getter_AddRefs(bodyContent));
-  NS_ENSURE_SUCCESS(rv, rv);
-  NS_ENSURE_ARG_POINTER(bodyContent);
+  NS_ENSURE_ARG_POINTER(aDoc);
 
-  uint32_t childCount = bodyContent->GetChildCount();
+  uint32_t childCount = aDoc->GetChildCount();
 
   // There are four possible range endpoints we might use:
   // DocumentStart, SelectionStart, SelectionEnd, DocumentEnd.
@@ -470,11 +435,11 @@
     offset = range->EndOffset();
 
     aSearchRange->SetStart(*node, offset, IgnoreErrors());
-    aSearchRange->SetEnd(*bodyContent, childCount, IgnoreErrors());
+    aSearchRange->SetEnd(*aDoc, childCount, IgnoreErrors());
     aStartPt->SetStart(*node, offset, IgnoreErrors());
     aStartPt->SetEnd(*node, offset, IgnoreErrors());
-    aEndPt->SetStart(*bodyContent, childCount, IgnoreErrors());
-    aEndPt->SetEnd(*bodyContent, childCount, IgnoreErrors());
+    aEndPt->SetStart(*aDoc, childCount, IgnoreErrors());
+    aEndPt->SetEnd(*aDoc, childCount, IgnoreErrors());
   }
   // Backward, not wrapping: DocStart to SelStart
   else if (mFindBackwards && !aWrap) {
@@ -488,12 +453,12 @@
     }
     offset = range->StartOffset();
 
-    aSearchRange->SetStart(*bodyContent, 0, IgnoreErrors());
-    aSearchRange->SetEnd(*bodyContent, childCount, IgnoreErrors());
-    aStartPt->SetStart(*node, offset, IgnoreErrors());
-    aStartPt->SetEnd(*node, offset, IgnoreErrors());
-    aEndPt->SetStart(*bodyContent, 0, IgnoreErrors());
-    aEndPt->SetEnd(*bodyContent, 0, IgnoreErrors());
+    aSearchRange->SetStart(*aDoc, 0, IgnoreErrors());
+    aSearchRange->SetEnd(*aDoc, childCount, IgnoreErrors());
+    aStartPt->SetStart(*aDoc, 0, IgnoreErrors());
+    aStartPt->SetEnd(*aDoc, 0, IgnoreErrors());
+    aEndPt->SetStart(*node, offset, IgnoreErrors());
+    aEndPt->SetEnd(*node, offset, IgnoreErrors());
   }
   // Forward, wrapping: DocStart to SelEnd
   else if (!mFindBackwards && aWrap) {
@@ -507,10 +472,10 @@
     }
     offset = range->EndOffset();
 
-    aSearchRange->SetStart(*bodyContent, 0, IgnoreErrors());
-    aSearchRange->SetEnd(*bodyContent, childCount, IgnoreErrors());
-    aStartPt->SetStart(*bodyContent, 0, IgnoreErrors());
-    aStartPt->SetEnd(*bodyContent, 0, IgnoreErrors());
+    aSearchRange->SetStart(*aDoc, 0, IgnoreErrors());
+    aSearchRange->SetEnd(*aDoc, childCount, IgnoreErrors());
+    aStartPt->SetStart(*aDoc, 0, IgnoreErrors());
+    aStartPt->SetEnd(*aDoc, 0, IgnoreErrors());
     aEndPt->SetStart(*node, offset, IgnoreErrors());
     aEndPt->SetEnd(*node, offset, IgnoreErrors());
   }
@@ -526,12 +491,12 @@
     }
     offset = range->StartOffset();
 
-    aSearchRange->SetStart(*bodyContent, 0, IgnoreErrors());
-    aSearchRange->SetEnd(*bodyContent, childCount, IgnoreErrors());
-    aStartPt->SetStart(*bodyContent, childCount, IgnoreErrors());
-    aStartPt->SetEnd(*bodyContent, childCount, IgnoreErrors());
-    aEndPt->SetStart(*node, offset, IgnoreErrors());
-    aEndPt->SetEnd(*node, offset, IgnoreErrors());
+    aSearchRange->SetStart(*aDoc, 0, IgnoreErrors());
+    aSearchRange->SetEnd(*aDoc, childCount, IgnoreErrors());
+    aStartPt->SetStart(*node, offset, IgnoreErrors());
+    aStartPt->SetEnd(*node, offset, IgnoreErrors());
+    aEndPt->SetStart(*aDoc, childCount, IgnoreErrors());
+    aEndPt->SetEnd(*aDoc, childCount, IgnoreErrors());
   }
   return NS_OK;
 }